MarkBind/markbind
# [WIP] Implement a Toasts component for MarkBind

**What is the purpose of this pull request?**




**Overview of changes:**

NOTE: After discussing with @yucheng11122017, this issue will be handed over to her, and will be resolved in PR #2237.

This Pull Request resolves #1960. It implements a Toasts component to MarkBind as per the requested specifications in the original issue:

- [X] New component declared using `<toast> ... </toast>` --> **Implemented as a new vue component.**
- [ ] Ideally a 10~20KB addition to our minified bundle at most --> **Needs to be re-investigated, but it was initially estimated to be around 20 ~ 25KB, which should still be acceptable.**
- [X] Supports multiple toasts in a single page --> **Multiple toasts on a single page, as well as toasts through sub-pages, are supported.**
- [X] Manually dismissible and/or timeout-able --> **Component is manually dismissable through clicking, and has the attribute `duration` that can be set manually (default is 10 seconds).**

Additionally, the list of things to be supported / are being supported:

- [X] Dynamic size based on the length of the message
- [X] Different color / style schemes supported in accordance with other components in MarkBind, e.g. [Badges](https://markbind.org/userGuide/components/presentation.html#badges)
- [ ] Custom styles supported via CSS attributed: `background`, `border`, `color`. --> **This may not be possible due to reliance on an external library, but can be trivially solved via the [Scoped CSS](https://vuejs.org/api/sfc-css-features.html#scoped-css) feature introduced in Vue 3, meaning that implementing custom styles may need to be on hold until migration from Vue 2 to Vue 3 for MarkBind.**

**Anything you'd like to highlight/discuss:**

Possible discussion points:
- Is there a need for a component that can manually call a toast on call / click (e.g. a button to create a toast)?
- Should the toast be opaque or translucent? (The original `alertify.js` library creates translucent toasts with `a = 0.95`.

**Testing instructions:**



Clone the code in this PR, and setup as per the instructions on the developer guide [here](https://markbind.org/devdocs/devGuide/development/settingUp.html#setting-up-the-dev-environment). Both `npm run setup` and `npm link` will need to be run for the new dependency introduced (`alertifyjs`).

Once set up, initialize a new MarkBind site / go to a pre-existing MarkBind site and add a new Toast component:

```html
<toast message="Testing." type="default" duration="10"></toast>
```

A toast should be displayed on the page that the toast was declared in. Note that `message` is a required attribute, while `type` and `duration` are not. The default value for these are `default` and `10`.

**Proposed commit message: (wrap lines at 72 characters)**
Implement a Toast component for MarkBind


---

**Checklist:** :ballot_box_with_check:
